### PR TITLE
Streetview dataset: Experiment 1

### PR DESCRIPTION
**Pull request by SwimmingPig**
This update includes a jupyter notebook that parse bounding boxes from tfrecords, and build tfrecord datasets.

The output dataset is a mixture of both the TC11 and UCF dataset.
Positive case: Only the largest confidence bbox which is larger than threshold will be cropped and considered as a positive case.
Negative case: Only if a picture does not contain a bbox that is greater than threshold will be considered.  (The image will not be cropped)

### CODE REVIEW

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    "tf.enable_eager_execution()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 21,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def read_TFRecord(file_path):\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: use all small case for function name: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#316-naming. Same to the other functions.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    "        'image/object/bbox/confidence': tf.io.VarLenFeature(tf.float32),\n",
+    "        'image/object/bbox/xmin': tf.io.VarLenFeature(tf.float32),\n",
+    "        'image/object/bbox/ymin': tf.io.VarLenFeature(tf.float32),\n",
+    "        'image/object/bbox/xmax': tf.io.VarLenFeature(tf.float32),\n",
+    "        'image/object/bbox/ymax': tf.io.VarLenFeature(tf.float32),\n",
+    "    }\n",
+    "\n",
+    "    def _parse_image_function(example_proto):\n",
+    "      # Parse the input tf.Example proto using the dictionary above.\n",
+    "      return tf.io.parse_single_example(example_proto, image_feature_description)\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Why do you need a nested function here? If it is necessary, make a comment explain the reason.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    "\n",
+    "    parsed_image_dataset = raw_image_dataset.map(_parse_image_function)\n",
+    "    return parsed_image_dataset\n",
+    "\n",
+    "def parse_labels(image_features):\n",
+    "    img_name = str(image_features['image/source_id'].numpy()).split('/')[-1][:-1]\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Make a comment explain the source_id string format, e.g. 'cns/path/to/image_file_name.jpg'

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    "    confidence = tf.sparse_tensor_to_dense(image_features['image/object/bbox/confidence'], default_value=0).numpy()\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Add a library to define those common const strings: 'image/object/bbox/confidence', etc.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    "    return img_name, confidence, bbox\n",
+    "\n",
+    "def img2example(img, label):\n",
+    "    imgByteArr = io.BytesIO()\n",
+    "    img.save(imgByteArr, format='JPEG')\n",
+    "    imgByteArr = imgByteArr.getvalue()\n",
+    "\n",
+    "    example = tf.train.Example(features=tf.train.Features(feature={\n",
+    "        \"image\": tf.train.Feature(bytes_list=tf.train.BytesList(value=[imgByteArr])),\n",
+    "        \"label\": tf.train.Feature(int64_list=tf.train.Int64List(value=[label]))}))\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb:110]: [optional] Follow the Lerna tf_example_keys conventions. Common tf keys in Lerna for reference (you may not need them all):

/****** Image properties. ******/
// kImageSourceId is commonly used in go/tf-example, while kImageId is commonly
// used in Geo.
inline constexpr char kImageId[] = "image/id";
inline constexpr char kImageSourceId[] = "image/source_id";
// Encoded image bytes, see supported formats in
// cs/symbol:lerna::common::image_formats.
inline constexpr char kImageData[] = "image/encoded";
inline constexpr char kImageFormat[] = "image/format";
// Image width and height of the image stored in tensorflow::Example.
inline constexpr char kImageWidth[] = "image/width";
inline constexpr char kImageHeight[] = "image/height";
// Number of channels in the image, for RGB it is 3, for grey scale it is 1.
inline constexpr char kImageChannels[] = "image/channels";

/****** Image level features. ******/
// Int64 labels.
inline constexpr char kImageClass[] = "image/class";
inline constexpr char kImageLabel[] = "image/class/label";
inline constexpr char kImageTextLabel[] = "image/class/text";
// Double labels
inline constexpr char kImageLabelConfidence[] = "image/class/confidence";
// This is used where the entire image has prediction labels.
inline constexpr char kImagePrediction[] = "image/class/prediction";
// Weight of each sample as a single element 1D tensor.
inline constexpr char kExampleWeight[] = "weight";

/****** Additional channel. ******/
inline constexpr char kImageAdditionalChannelsData[] =
    "image/additional_channels/encoded";
inline constexpr char kImageAdditionalChannelsFormat[] =
    "image/additional_channels/format";

/****** For box detection. ******/
// Bounding box for each object. These are normalized to unit image width and
// height dimension.
inline constexpr char kBboxXmin[] = "image/object/bbox/xmin";
inline constexpr char kBboxXmax[] = "image/object/bbox/xmax";
inline constexpr char kBboxYmin[] = "image/object/bbox/ymin";
inline constexpr char kBboxYmax[] = "image/object/bbox/ymax";
// Bounding box "objectness" confidence i.e. confidence bounding box holds an
//  object. This should be one float per bounding box.
inline constexpr char kBboxConfidence[] = "image/object/bbox/confidence";
// An integer of box label.
inline constexpr char kBboxLabel[] = "image/object/class/label";

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    "\n",
+    "                if label:\n",
+    "                    img = img.crop(bbox)\n",
+    "                img = img.resize((64, 64))\n",
+    "\n",
+    "                example = img2example(img, label)\n",
+    "                writer.write(example.SerializeToString())\n",
+    "            except Exception as e:\n",
+    "                print(e)\n",
+    "                print(img_path + \" is not valid\")\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Avoid using try-exception clause if possible: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#24-exceptions. If you have to use it, minimize the try block.

In this case, if Image.open can return a success or failure status, use if condition instead. Otherwise, only put Image.open() in the try block.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    "            try:\n",
+    "                img = Image.open(img_path, \"r\")\n",
+    "\n",
+    "                # Exclude all non RGB images\n",
+    "                if len(img.getbands()) != 3:\n",
+    "                    continue\n",
+    "\n",
+    "                if label:\n",
+    "                    img = img.crop(bbox)\n",
+    "                img = img.resize((64, 64))\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Make global const variables for const numbers, or make image size as a configurable param.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+    "\n",
+    "    example = tf.train.Example(features=tf.train.Features(feature={\n",
+    "        \"image\": tf.train.Feature(bytes_list=tf.train.BytesList(value=[imgByteArr])),\n",
+    "        \"label\": tf.train.Feature(int64_list=tf.train.Int64List(value=[label]))}))\n",
+    "    \n",
+    "    return example"
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb:124]: Why not put all function definitions in one cell?

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 1 above)
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Add a function description to each function unless it is trivial.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def write_TFRecord_UCF(parsed_image_dataset, folder_path, writer):\n",
+    "    for image_features in parsed_image_dataset:\n",
+    "        img_name, confidence, bbox = parse_labels(image_features)\n",
+    "\n",
+    "        if img_name:\n",
+    "            label = 0\n",
+    "            if confidence.size>0 and max(confidence)>0.9:\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Make the confidence thresholds as params and define global const variables for them.

Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: [nit] add space around "<" and ">"

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 10]
```diff
@@ -0,0 +1,221 @@
+{
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb:1]: Add a file description at the top.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# input tfrecord file path\n",
+    "input_record_path = './streetlearn-detections/'\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Use CAPS_WITH_UNDER for global constants: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#3164-guidelines-derived-from-guidos-recommendations

Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: For file dirs or file names, add "input_" or "output_" to indicate whether it is input or output for better readability.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Write train and test tfrecord for TC11 dataset\n",
+    "train_image_path = os.path.join(TC11_img_dir, 'train')\n",
+    "test_image_path = os.path.join(TC11_img_dir, 'test')\n",
+    "\n",
+    "write_TFRecord_TC11(train_image_path, train_writer)\n",
+    "write_TFRecord_TC11(test_image_path, test_writer)"
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: I am confused with this part. Do you write TC11 dataset to both train set and test set?

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Write train and test tfrecords for UCF dataset\n",
+    "for i in range(900):\n",
+    "    file_name = \"./streetlearn_detections_tfexample-\" + str(i).zfill(5) + \"-of-01000.tfrecord\"\n",
+    "    parsed_image_dataset = read_TFRecord(os.path.join(input_record_path, file_name))\n",
+    "    write_TFRecord_UCF(parsed_image_dataset, UCF_img_dir, train_writer)\n",
+    "\n",
+    "for i in range(900,1000):\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Add a global constant variable for 1000. Use 0.9*1000 to get the train/test split or add another global constant for 900.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 2 above)
Comment (yuxizhang-google) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: nvm, I realized you pass this function as a param in the next line.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 12 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: This PR was created last week, at that time TC11 is written into both train and test set for this version 1 dataset.
Do you want me to update to a newer version? Or keep this version for this PR and the updated other versions of the dataset will be created in future PRs.

For me, I think it's better to keep this PR as it is right now, cause future versions of the dataset don't only change this part of the code. But it's up to you for the final decision.

Comment (yuxizhang-google) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: That makes sense to me. It's fine to keep this version in this PR.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 1 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done!

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 3 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 4 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: I've added a global constant for these.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 5 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb:110]: Great info. But I'll skip this part right now to maintain backward compatibility of the model. If I need more labels later on, I'll follow the naming scheme.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 10 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb:1]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 11 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 13 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 6 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Only Image.open() is in the try-exception clause right now.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 7 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 8 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb:124]: The later part is specific for streetview_v1 dataset. So I want to separate general functions and dataset specific functions. I've added some notes on the notebook file to make it more distinct.

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 1 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb (hunk 9 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

File: Mixmatch/streetview_dataset/parse_TFRecord.ipynb [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Write train and test tfrecords for UCF dataset\n",
+    "for i in range(0.9*INPUT_RECORD_CNT):\n",
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: nit: add spaces around *, same below.

Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/parse_TFRecord.ipynb]: Done

**Review by yuxizhang-google — APPROVED** [review]
